File: src/001.Array/697.DegreeOfAnArray/697.数组的度.js
/*
 * @lc app=leetcode.cn id=697 lang=javascript
 *
 * [697] 数组的度
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @return {number}
 */
var findShortestSubArray = function (nums) {
	const numsObj = {};
	let maxLen = 0;
	let degree = nums.length;

	for (const i in nums) {
		numsObj[nums[i]] = [...(numsObj[nums[i]] || []), Number(i)];
	}

	for (const i in numsObj) {
		const indexs = numsObj[i];
		const tempLen = indexs[indexs.length - 1] - indexs[0] + 1;

		if (indexs.length > maxLen) {
			maxLen = indexs.length;
			degree = tempLen;
		} else if (indexs.length == maxLen && tempLen < degree) {
			maxLen = indexs.length;
			degree = tempLen;
		}
	}

	return degree;
};
// @lc code=end

File: src/001.Array/697.DegreeOfAnArray/697.数组的度.go
/*
 * @lc app=leetcode.cn id=697 lang=golang
 *
 * [697] 数组的度
 */

// @lc code=start

type entry struct{ cnt, l, r int }

func findShortestSubArray(nums []int) int {
	mp := map[int]entry{}

	for i, v := range nums {
		if e, has := mp[v]; has {
			e.cnt++
			e.r = i
			mp[v] = e
		} else {
			mp[v] = entry{1, i, i}
		}
	}

	maxCnt := 0
	degree := 0
	for _, e := range mp {
		if e.cnt > maxCnt {
			maxCnt, degree = e.cnt, e.r-e.l+1
		} else if e.cnt == maxCnt {
			degree = min(degree, e.r-e.l+1)
		}
	}

	return degree
}

func min(a, b int) int {
	if a > b {
		return b
	}

	return a
}

// @lc code=end

